src/v/wasm/transform_module_v2.cc

```diff
@@ -0,0 +1,227 @@
+#include "wasm/wasi.h"
+
+#include <seastar/core/condition-variable.hh>
+#include <seastar/coroutine/maybe_yield.hh>
+
+#include <exception>
+#include <optional>
+
+namespace wasm {
+
+constexpr int32_t NO_ACTIVE_TRANSFORM = -1;
+constexpr int32_t INVALID_BUFFER = -2;
+
+transform_module_v2::transform_module_v2(wasi::preview1_module* m)
+  : _wasi_module(m) {}
+
+ss::future<ss::chunked_fifo<model::transformed_data>>
+transform_module_v2::for_each_record_async(
+  model::record_batch input, ss::noncopyable_function<void()> cb) {
+    vassert(
+      input.header().attrs.compression() == model::compression::none,
+      "wasm transforms expect uncompressed batches");
+
+    iobuf_const_parser parser(input.data());
+
+    ss::chunked_fifo<record_metadata> records;
+    records.reserve(input.record_count());
+    size_t max_size = 0;
+
+    while (parser.bytes_left() > 0) {
+        auto [record_size, rs_amt] = parser.read_varlong();
+        auto attrs = parser.consume_type<model::record_attributes::type>();
+        auto [timestamp_delta, td_amt] = parser.read_varlong();
+        auto [offset_delta, od_amt] = parser.read_varlong();
+        size_t meta_size = sizeof(decltype(attrs)) + td_amt + od_amt;
+        size_t payload_size = record_size - meta_size;
+        max_size = std::max(payload_size, max_size);
+        parser.skip(payload_size);
+        model::timestamp ts = input.header().max_timestamp;
+        if (
+          input.header().attrs.timestamp_type()
+          == model::timestamp_type::create_time) {
+            ts = model::timestamp(
+              input.header().first_timestamp() + timestamp_delta);
+        }
+        records.push_back({
+          .metadata_size = rs_amt + meta_size,
+          .payload_size = payload_size,
+          .attributes = model::record_attributes(attrs),
+          .timestamp = ts,
+          .offset = input.base_offset() + offset_delta,
+        });
+    }
+
+    _call_ctx.emplace(batch_transform_context{
+      .batch_header = input.header(),
+      .batch_data = std::move(input).release_data(),
+      .max_input_record_size = max_size,
+      .records = std::move(records),
+      .output_data = {},
+      .record_callback = std::move(cb),
+    });
+
+    co_await host_wait_for_proccessing();
+
+    auto result = std::exchange(_call_ctx, std::nullopt);
+    co_return std::move(result->output_data);
+}
+
+void transform_module_v2::check_abi_version_1() {
+    // This function does nothing at runtime, it's only an opportunity for
+    // static analysis of the module to determine which ABI version to use.
+}
+
+// NOLINTBEGIN(bugprone-easily-swappable-parameters)
+ss::future<int32_t> transform_module_v2::read_batch_header(
+  int64_t* base_offset,
+  int32_t* record_count,
+  int32_t* partition_leader_epoch,
+  int16_t* attributes,
+  int32_t* last_offset_delta,
+  int64_t* base_timestamp,
+  int64_t* max_timestamp,
+  int64_t* producer_id,
+  int16_t* producer_epoch,
+  int32_t* base_sequence) {
```

Comment:
🥇

Reply:
Haha we don't expose all this information right now, so I'm kind of tempted to omit it... 🤷